internal/genericpubsub/genericpubsub.go

```diff
@@ -0,0 +1,133 @@
+/*
+ *
+ * Copyright 2023 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package genericpubsub provides functionality to report and track
+// target changes of struct which is registered as a subscriber.
+package genericpubsub
+
+import (
+	"context"
+	"sync"
+
+	"google.golang.org/grpc/internal/grpcsync"
+)
+
+// Watcher wraps the functionality to be implemented by components
+// interested in watching target changes.
+type Watcher interface {
+	// OnTargetChange is invoked to report target changes on the
+	// entity being watched.
+	OnTargetChange(target interface{})
+}
+
+// Tracker provides pubsub-like functionality for target changes.
+//
+// The entity whose target is being tracked publishes updates by
+// calling the SetTarget() method.
+//
+// Components interested in target updates of the tracked entity
+// subscribe to updates by calling the AddWatcher() method.
+type Tracker struct {
```

Comment:
Based on an offline discussion, we think it would be nice to have the following API. 

The implementation can remain mostly unchanged though. But we should replace `target` with `message` as the latter makes more sense for a generic pubsub utility.

```golang
package grpcsyc

// PubSub provides pubsub like functionality for arbitrary messages.
//
// It guarantees that published messages will be delivered to subscribers in the same 
// order in which they were published. It also guarantees lossless delivery of published
// messages to subscribers until Stop() is called, after which no messages will be delivered.
type PubSub { ... } // <<< this is equivalent to the Tracker type that we have >>>

// Creates a new pubsub.
func NewPubSub() *PubSub { ... } <<< This function will not accept any initial message >>>

// Publish a new message.
// Returns nothing. If called after `Stop()`, this will be no-op.
func (ps *PubSub) Publish(msg interface{}) { ... } // <<< This is equivalent to SetTarget >>>

// A convenience type to represent the callback.
type OnChange func(interface{}) // this is equivalent to the Watcher interface, which is a single method interface and can be replaced with a user supplied callback instead.

// API to subscribe to changes. 
// Provides a guarantee that the callback will not be invoked after `Stop()` returns.
// Returned cancel func is to be invoked by the caller to cancel the subscription.
func (ps *PubSub) Subscribe(f OnChange) (cancel func()) { ... }

// Once stopped, new messages cannot be published and existing subscribers' callback
// will no longer be invoked
func (ps *PubSub) Stop() { ... }
```

Reply:
@easwars : I have a question and opinion.

Question:
If `NewPubSub()` does not accept any initial messages, then the `Schedule` at `<1>` on the following is not necessary or should be skipped in case the `PubSub`'s `msg` is nil?
```go
func (ps *PubSub) Subscribe(watcher Watcher) func() {
	ps.mu.Lock()
	defer ps.mu.Unlock()

	if ps.stopped {
		return func() {}
	}

	ps.watchers[watcher] = true

	msg := ps.msg

        // <1>
	ps.cs.Schedule(func(context.Context) {
		ps.mu.Lock()
		defer ps.mu.Unlock()
		watcher.OnChange(msg)
	})

	return func() {
		ps.mu.Lock()
		defer ps.mu.Unlock()
                // <2>
		delete(ps.watchers, watcher)
	}
}
```

Opinion:
If we adopt the use of `OnChange` like the following, how can we delete a specified `func(interface{})` from the array of `func(interface{})`? (at `<2>` on the above)
Also, I think that the implementation with using `Watcher` struct is simple enough.
Could you please tell me reasons to use `type OnChange func(interface{})`?
```go
// A convenience type to represent the callback.
type OnChange func(interface{}) // this is equivalent to the Watcher interface, which is a single method interface and can be replaced with a user supplied callback instead.
```

Reply:
> Question1:
> If NewPubSub() does not accept any initial messages, then the Schedule at <1> on the following is not necessary or > should be skipped in case the PubSub's msg is nil?

Yes, the `Schedule` at the location pointed by you should be skipped if `ps.msg == nil`.

> Question2:
> If we adopt the use of OnChange like the following, how can we delete a specified func(interface{}) from the array of > func(interface{})? (at <2> on the above)

Actually, you are correct. Having an interface with a single method in it does enable storing them in a map and deleting them from the map when the cancel func is invoked. Functions are not a comparable type and hence cannot be used as map keys.

So, let's go with the `Watcher` interface, but with other changes I suggested. Thank you.